3주차: ToDo REST API 테스트 코드 작성하기

steve7867 opened the pull request:
안녕하세요!
벌써 3주차네요.
제일 기대했던 테스트 코드 작성 과제네요.
이번 주도 잘 부탁드립니다!!

steve7867 committed:
TaskController 유닛 테스트 작성

steve7867 committed:
Task 모델 테스트 작성

steve7867 committed:
TaskControllerTest 예외 메시지 검증 로직 추가, TaskServiceTest 추가

steve7867 commented:
오늘은 TaskController 유닛 테스트, Task 모델 테스트까지만 테스트 코드 작성을 마쳤습니다.

HelloController는 없는 것 같은데 HelloController에 대한 테스트 코드는 작성하지 않아도 되나요?

ssisksl77 commented:
> 오늘은 TaskController 유닛 테스트, Task 모델 테스트까지만 테스트 코드 작성을 마쳤습니다.
> 
> HelloController는 없는 것 같은데 HelloController에 대한 테스트 코드는 작성하지 않아도 되나요?

네 HelloController 는 무시해도 됩니다!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,47 @@
+        }
+    }
+
+}
```
nit : EOF 에는 빈칸을 추가해주세요!

ssisksl77 replied:
너무 귀찮다면 intellij의 기능을 한번 사용해보세요! 시프트를 두번 누르고 Ensure every saved file ends 를 검색해보세요!

steve7867 replied:
파일 마지막에 개행 문자를 넣는 게 유닉스에서 관용적으로 사용되어 오던 것이군요.
간혹 파일 마지막에 개행 문자가 없으면 올바르게 처리하지 못하는 프로그램도 있다고 하네요.
물론 요즘 IDE에서는 문제가 되지는 않지만 관용이기에 마지막 줄 개행을 하는 것이 더 안정적일 것 같습니다.
github에서도 마지막 줄에 개행 문자를 넣지 않으면 경고 아이콘이 뜨네요.

Intellij 설정도 했습니다. 감사합니다~

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,212 @@
+        @DisplayName("get all tasks")
+        void list() {
+            assertThat(controller.list()).isEmpty();
+        }
```
현재 테스트하고 있는 내용을 DisplayName에 녹여보는 것이 시작이겠습니다!

steve7867 replied:
테스트 코드를 계층 구조로 변경했습니다.
Describe-Context-It 패턴으로 테스트 코드를 작성했습니다.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,212 @@
+
+        @Test
+        @DisplayName("post a task")
+        void create() {
+            final int oldSize = controller.list().size();
+
+            final Task src = new Task();
+            src.setTitle(TASK_TITLE);
+            assertThatNoException().isThrownBy(() -> controller.create(src));
+
+            final List<Task> taskList = controller.list();
+
+            final int newSize = taskList.size();
+            assertThat(newSize - oldSize).isEqualTo(1);
+
+            final Task addedTask = taskList.get(0);
+            assertThat(addedTask).isNotEqualTo(src);
+
+            final Long addedTaskId = addedTask.getId();
+            assertThat(addedTaskId).isNotNull();
+            assertThat(addedTaskId).isGreaterThan(0L);
```
여기도요! Task 등록을 테스트하는 것은 알겠지만!

좀 더 상세하게 테스트를 설명했으면 좋겠으며! assertThat과 assertThatNoException 을 보았을 때는 테스트를 쪼개보는 것도 좋을 것 같아요!

steve7867 replied:
테스트 목적에 따라 메소드를 나눴습니다

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,47 @@
+        @Test
+        @DisplayName("get a task")
+        void getTask() {
+            final Long randomId = getRandomId();
```
getRandomId 를 쓸 필요가 있을까요? 

만약 있다고 생각하신다면 댓글을 남겨보아요!

steve7867 replied:
생각해보니까 굳이 쓸 필요는 없을 것 같네요.
그냥 제가 임의의 수를 하나 정해서 사용해도 괜찮을 것 같습니다.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,212 @@
+        @DisplayName("update a task with valid id")
+        void updateWithValidId() {
+            final Task task = new Task();
+            final String newTitle = TASK_TITLE + "`";
```
`"`"` 는 별 뜻 없는 상수값인거죠? ㅎㅎ

steve7867 replied:
저렇게 사용하는 것보다 앞으로는 변수를 하나 선언해서 사용하도록 해야겠습니다.
코드 가독성도 안 좋고 실수의 여지도 있어 보이네요.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,212 @@
+            return invalidIdOptional.orElse(Long.MAX_VALUE);
+        }
+    }
+}
```
여기도 EOF에 빈칸을 넣어주세요!

steve7867 replied:
추가했습니다.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,212 @@
+        }
+
+        Long getInvalidId() {
+            final Long validId = getValidId();
+
+            Optional<Long> invalidIdOptional = new Random().longs(100).boxed()
+                    .filter(l -> !l.equals(validId))
+                    .findAny();
+
+            return invalidIdOptional.orElse(Long.MAX_VALUE);
+        }
```
getValidId() 는 valid한 Id 중에 하나만 갖고 있는데 `get.(0)` 이것으로 InValid 함을 테스트해도 괜찮은건가요?

또한 스트림까지 만들어서 안맞으면 `Long.MAX_VALUE` 로 리턴하는 것인데 그냥 처음부터 `Long.MAX_VALUE` 를 리턴하는 것보다 어떤면이 좋아진 것인지 설명해보시면 좋을 것 같아요!

steve7867 replied:
불필요한 코드를 너무 많이 넣은 것 같네요.
Long.MAX_VALUE를 invalid한 id로 사용하도록 수정했습니다.

steve7867 committed:
TaskServiceTest 계층 구조로 변경, TaskGenerator 추가

steve7867 committed:
TaskServiceTest 버그 수정, TaskNotFoundException 위치 변경

TaskServiceTest 버그 수정: addedTask의 title과 src의 title이 같은 경우를 방지하는 로직 추가
TaskNotFoundException 위치 변경: exceptions/로 이동

steve7867 committed:
TaskControllerTest 계층 구조로 변경, TaskServiceTest custom display name 수정

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/TaskGenerator.java:
```diff
@@ -0,0 +1,26 @@
+        return task;
+    }
+
+    public static Task generateTaskWithRandomTitle() {
+        return generateTask(getRandomTitle());
+    }
```
ㅎㅎ 재밌군요!! 주석을 추가해주세요! 
이런류의 메소드에는 예시도 함께 있으면 도움이 많이됩니다!

steve7867 replied:
네 작성했습니다.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/TaskGenerator.java:
```diff
@@ -0,0 +1,26 @@
+public class TaskGenerator {
+
+    public static Task generateTask(String title) {
+        final Task task = new Task();
+        task.setTitle(title);
+        return task;
```
이펙티브자바에 item2 를 읽어보고 이 패턴이 무엇인지 알아보아요!

steve7867 replied:
자바 빈즈 패턴이네요.
자바 빈즈 패턴의 경우 가독성은 나쁘지 않지만, 모든 필드의 값을 세팅하기 전까지 객체의 일관성이 무너져 있다는 문제점이 있습니다.
빌더 패턴 사용을 고려해봤는데 필드가 id, title 2개 뿐이라 빌더 패턴을 사용하면 오히려 코드만 길어지는 단점이 있는 것 같습니다.
그래서 Task를 불변 객체로 만들고 id, title에 대한 매개변수를 모두 받는 생성자를 만드는 방법으로 Task 클래스를 수정했습니다.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,297 @@
+            void it_throws_a_TaskNotFoundException() {
+                addOneTask();
+                final Long id = Long.MAX_VALUE;
+                final Task task = TaskGenerator.generateTaskWithRandomTitle();
```
지금도 아주 완벽하지만 BDD 스타일을 조금만 더 맛보자면! Context에 해당하는 정보들은 이곳에 있으면 안됩니다!

steve7867 replied:
Context에 해당하는 코드들은 Context 클래스의 필드나 `@BeforeEach` 메소드로 뺐습니다.

steve7867 committed:
Task: 불변 객체화, TaskService: List -> Map으로 변경

steve7867 committed:
RandomTitleGenerator 생성, TaskService: taskExists() 수정

steve7867 committed:
TaskServiceTest 리팩토링

테스트 코드를 BDD 패턴에 더 부합하도록 수정하여 가독성 향상

steve7867 committed:
TaskServiceTest 버그 수정, NumberGenerator와 RandomTitleGenerator 생성

TaskServiceTest 버그 수정: @Nested 클래스의 인스턴스 메소드에 적용된 @BeforeAll을 @BeforeEach로 변경
NumberGenerator와 RandomTitleGenerator 생성: 테스트에서 사용하는 숫자와 문자열을 생성하는 책임을 NumberGenerator와 RandomTitleGenerator에 위임

steve7867 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,356 @@
+
+                    assertThat(task.getId()).isEqualTo(idHavingMappedTask);
+                }
+            }
```
해당 테스트를 실행하는 경우 상위 클래스에 있는 `@BeforeEach` 메소드가 실행이 안 되네요.
Junit5 guide 문서에 테스트 인스턴스 생명 주기 부분과 `@Nested` class 부분을 읽어 보고 구글링 해도 뭐가 문제인지 찾을 수가 없네요.
디버깅 해보니 최상위 클래스의 인스턴스와 `@Nested`가 붙은 클래스의 인스턴스들은 모두 생성되는데 `@BeforeEach` 메소드들만 실행이 안 되네요.

137번 라인을
```java
private final Long idHavingMappedTask = 1L;
```
로 바꾸니까 상위 클래스들의 `@BeforeEach` 메소드가 정상적으로 작동하는데 도대체 뭐가 문제인가요?

steve7867 replied:
해결했습니다.
Junit은 제가 작성한 테스트 코드처럼 `@Nested` 클래스를 이용한 계층적 구조를 가질 때, Outermost class -> Innermost class 순서로 객체의 생성과 초기화를 끝낸 후, Outermost class -> Innermost class 순서로 `@BeforeEach` 메소드를 실행하네요. 이거 해결한다고 시간 많이 허비했네요 ㅎㅎ

steve7867 replied:
> [1d224e0](https://github.com/CodeSoom/spring-week3-assignment-1/commit/1d224e021f0acbbe5dc2ae4fba860c7a9e540a4a) 커밋의 내용이 좀 더 구체적이었으면 좋겠어요! '어떤 문제를 무엇으로 해결되었다'

커밋 메시지가 명확하지 않네요.
해당 커밋에서 해결한 문제는 이 문제입니다.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/utils/NumberGenerator.java:
```diff
@@ -0,0 +1,14 @@
+import java.util.Random;
+
+public class NumberGenerator {
+    public static int getRandomIntegerBetweenOneAndOneHundred() {
+        return new Random().nextInt(101);
+    }
```
좋습니다. 

`new Random().nextInt(101)` 을 스태틱 메소드인 `getRandomIntegerBetweenOneAndOneHundred` 분리하셨습니다!

본인이 생각하는 분리해서 얻는 장점에 대해서 공유해주세요!

steve7867 replied:
1. 코드를 메소드로 분리하게 되면 메소드명만 보고도 코드가 어떤 역할을 하는지 파악하기 쉽습니다.
2. 여러번 반복되어 사용되는 동일한 코드를 일일히 타이핑하지 않고 메소드 호출 하나만으로 끝낼 수 있죠.
3. 2번과 관련있는 것인데, 만약 코드의 로직을 고쳐야 할 일이 있을 때 동일한 코드들을 메소드를 사용하지 않은 경우에는 여러 곳의 코드를 일일히 수정해야 하는 불편함이 있습니다. 실수할 여지가 많은 건 두말할 필요도 없죠. 하지만 메소드를 사용하면 메소드의 코드만 고치면 간단하게 수정 작업을 끝낼 수 있습니다.

테스트 코드의 경우 반복되는 코드가 유난히 많고 여러가지 케이스를 다 커버하려니까 코드도 많이 길어집니다.
반복되는 코드들은 최대한 메소드로 분리하여 코드량을 줄여야겠습니다.

ssisksl77 replied:
@steve7867 여기에 코멘트를 남기신 걸까요? 혹시 pending 상태는 아닌지 확인부탁드려요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/TaskService.java:
```diff
@@ -4,51 +4,64 @@
 
-    public List<Task> getTasks() {
-        return tasks;
+    public Collection<Task> getTasks() {
+        return Collections.unmodifiableCollection(taskMap.values());
```
좋아요! Collection과 List의 차이를 공유해보세요!

ssisksl77 replied:
unmodifiableCollection !! 아주 좋아요! 이것도 한번 코멘트로 공유해주세요!

아참! 참고자료도 함께요! 
이왕이면 공식문서에 가까운 정보일수록 좋습니다!

steve7867 replied:
Collection의 경우 List, Set, Queue의 상위 인터페이스로써 Map을 구현하는 클래스를 제외한 나머지 컬렉션 프레임워크 클래스들의 최상위 인터페이스라고 할 수 있습니다. Collection의 경우 요소의 중복을 허용할 수도 있고 허용하지 않을 수도 있고, 요소의 순서를 보장할 수도 있고 보장하지 않을 수도 있습니다. List의 경우, 요소의 중복을 허용하고 순서를 보장한다는 점에서 차이가 있습니다. 반면 Set의 경우에는 중복을 불허하고 순서도 보장하지 않습니다. 물론 LinkedHashSet의 경우는 중복은 불허하지만 루핑할 때 순서를 보장해주는 Set도 있습니다.

steve7867 replied:
unmodifiableCollection 은 Collection에 대한 view를 제공하는 메소드라고 볼 수 있습니다. unmodifiableCollection 메소드를 통해 반환받은 컬렉션에 대해서는 mutator method가 허용되지 않고 read-only 접근만 허용되는 특성이 있습니다. view를 통해 original collection이 변형되는 것을 막기 위함이죠.

ssisksl77 replied:
@steve7867 여기에 코멘트를 남기신 걸까요? 혹시 pending 상태는 아닌지 확인부탁드려요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/TaskService.java:
```diff
@@ -4,51 +4,64 @@
+            throw new TaskNotFoundException(id);
+        }
+
+        return taskMap.get(id);
```
`taskMap.get(i)` 을 두번하는 구조군요? 한번만 할 수는 없을까요?

steve7867 replied:
한번만 하도록 개선했습니다.

ssisksl77 replied:
@steve7867 여기에 코멘트를 남기신 걸까요? 혹시 pending 상태는 아닌지 확인부탁드려요!

steve7867 committed:
TaskGenerator 삭제, NumberGenerator 주석 추가

steve7867 committed:
TaskControllerTest, TaskServiceTest 버그 수정

`@Nested` 클래스 내에서 필드 초기화 부분을 `@BeforeEach` 메소드로 처리하도록 변경

steve7867 committed:
NoneTaskRegisteredException, NullTaskException, NegativeIdException, InvalidTaskTitleException 추가

steve7867 committed:
IdValidator, TaskValidator 추가

steve7867 committed:
TaskControllerTest 테스트 추가 및 수정

steve7867 committed:
TaskController 기능 구현

TaskControllerTest의 모든 테스트 코드를 통과하도록 기능 구현

steve7867 committed:
TaskControllerTest, TaskServiceTest: custom display name 변경, TaskTest: BDD 패턴으로 변경

steve7867 committed:
TaskControllerTest: 코드 리팩토링, TaskService: task 저장소 변경

TaskService task 저장소 변경: HashMap에서 TreeMap으로 변경. task를 추가된 순서대로 저장하기 위함.

steve7867 committed:
TaskControllerMvcTest 생성

steve7867 committed:
TaskControllerMvcTest 테스트 코드 추가

steve7867 committed:
TaskControllerMvcTest create() 테스트 코드 추가

steve7867 committed:
TaskService 버그 수정

taskMap에 대한 중복적인 접근을 제거

ssisksl77 commented:
안녕하세요! 
많은 커밋을 만드셨군요! 

전날 적어놓은 코멘트를 읽어보시고 코멘트를 달아주세요!

요구사항을 해결하는 것도 좋지만 찬찬히 생각해보는 시간을 갖는 것도 중요할거에요!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerMvcTest.java:
```diff
@@ -62,7 +65,7 @@ void setUp() {
                 .build();
 
-        registeredTaskList = new ArrayList<>();
+        registeredTaskMap = new TreeMap<>();
```
오!! 좋습니다 HashMap을 쓰지않고 TreeMap을 쓴 이유가 무엇인가요?!

steve7867 replied:
registeredTaskMap.values() 메소드를 호출하면 task가 생성된 순서대로 컬렉션에 담겨서 반환되기 때문입니다.

ssisksl77 replied:
@steve7867 여기에 코멘트를 남기신 걸까요? 혹시 pending 상태는 아닌지 확인부탁드려요!

ssisksl77 commented:
1d224e021f0acbbe5dc2ae4fba860c7a9e540a4a 커밋의 내용이 좀 더 구체적이었으면 좋겠어요!
'어떤 문제를 무엇으로 해결되었다'

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/exceptions/NegativeIdException.java:
```diff
@@ -0,0 +1,10 @@
+    private static final String MESSAGE_POSTFIX = ": Id는 0 또는 양수만 허용됩니다.";
+
+    public NegativeIdException(Long id) {
+        super(id + MESSAGE_POSTFIX);
```
위에 있는 `InvalidTaskTitleException` 은 인라인인데 여기는 상수를 만드셨군요!

둘의 차이는 무엇인가요?

steve7867 replied:
수정했습니다.
코드 스타일을 일관적으로 유지해야겠습니다.

ssisksl77 replied:
@steve7867 여기에 코멘트를 남기신 걸까요? 혹시 pending 상태는 아닌지 확인부탁드려요!

steve7867 replied:
죄송해요
제가 깃헙 사용에 안 익숙해서요ㅠㅠ

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -2,39 +2,48 @@
     @PutMapping("{id}")
     public Task update(@PathVariable Long id, @RequestBody Task task) {
+        IdValidator.validate(id);
+        TaskValidator.validate(task);
```
여기서 task는 `id` 가 null 인가요?

steve7867 replied:
네 맞습니다.

steve7867 commented:
> 안녕하세요! 많은 커밋을 만드셨군요!
> 
> 전날 적어놓은 코멘트를 읽어보시고 코멘트를 달아주세요!
> 
> 요구사항을 해결하는 것도 좋지만 찬찬히 생각해보는 시간을 갖는 것도 중요할거에요!

코멘트 남겼습니다~

steve7867 committed:
InvalidTaskTitleException, NegativeIdException, NullTaskException 모두 IllegalArgumentException를 상속하도록 변경

steve7867 committed:
TaskErrorAdvice: @RestControllerAdvice로 변경, handleIllegalArgument 메소드 추가

handleIllegalArgument: NegativeIdException, NullTaskException, InvalidTaskTitleException에 대한 예외 처리

steve7867 committed:
TaskController patch 메소드 제거, TaskService deleteTask 리턴 타입 변경

TaskController patch 메소드 제거: update와 patch가 같은 기능을 하는 메소드이므로 update 메소드로 병합
TaskService deleteTask 리턴 타입 변경: 리턴 타입이 void인 경우 Mockito 사용이 불가하므로 리턴 타입을 Task로 변경

steve7867 committed:
TaskControllerMvcTest 테스트 코드 추가

TaskController의 update, delete 메소드에 대한 테스트 코드 추가

steve7867 committed:
RandomTitleGenerator: getRandomTitleDifferentFrom 메소드 추가, Task: equals 메소드 추가

steve7867 committed:
불필요한 필드 삭제, custom display name 일부 변경

steve7867 committed:
addRandomNumberOfTasks 메소드 추가, 일부 필드명 간소화

steve7867 committed:
Response Body 검증 로직 수정

steve7867 committed:
TaskTest: equals 메소드 테스트 코드 추가, TaskServiceTest: 공백 라인 제거

steve7867 commented:
테스트 코드 작성을 모두 마쳤는데, 궁금한 게 있습니다.

1. 테스트 코드량이 너무 많은 것 같습니다. TaskController, TaskService의 코드 라인은 60 줄 정도인데 두 클래스에 대한 테스트 코드는 500, 300 줄이 넘어갑니다. 제가 테스트 코드를 제대로 작성한 게 맞는지 궁금합니다.

2. TaskController 유닛 테스트에 대한 질문입니다. TaskController의 경우 TaskService에 의존하는 클래스입니다. 그럼 TaskController에 대한 유닛 테스트를 진행할 때 TaskService 객체는 MockBean으로 주입받아서 테스트해야 하는 건지 아니면 진짜 TaskService 객체를 생성해서 테스트해야 하는 건지 판단이 서질 않습니다.
(현재는 진짜 TaskService 객체를 생성해서 테스트했습니다.)

3. 테스트를 할 때 사용하는 예외 클래스나 util 클래스의 경우, test 디렉토리 아래에 생성해도 괜찮나요? 아니면 main 디렉토리에 생성하는 것이 올바른 것인가요?

steve7867 committed:
NumberGenerator: getRandomNotNegativeLong 내 변수명 수정, TaskControllerTest: 공백 라인 제거

ssisksl77 commented:
> 테스트 코드 작성을 모두 마쳤는데, 궁금한 게 있습니다.
> 
> 1. 테스트 코드량이 너무 많은 것 같습니다. TaskController, TaskService의 코드 라인은 60 줄 정도인데 두 클래스에 대한 테스트 코드는 500, 300 줄이 넘어갑니다. 제가 테스트 코드를 제대로 작성한 게 맞는지 궁금합니다.
> 2. TaskController 유닛 테스트에 대한 질문입니다. TaskController의 경우 TaskService에 의존하는 클래스입니다. 그럼 TaskController에 대한 유닛 테스트를 진행할 때 TaskService 객체는 MockBean으로 주입받아서 테스트해야 하는 건지 아니면 진짜 TaskService 객체를 생성해서 테스트해야 하는 건지 판단이 서질 않습니다.
>    (현재는 진짜 TaskService 객체를 생성해서 테스트했습니다.)


1. 
하나의 테스트 (assert 구문) 을 만들기 위한 보일러플레이트가 많기 때문이겠죠.
하지만 테스트를 만들다보면 하나의 기능을 테스트하는데 여러가지의 테스트를 하는 것은 자연스럽습니다!

2. 저는 가능하면 Mock을 사용하지 않으려고 노력합니다. 차라리 FakeObject 만들기를 선호하죠! 하지만 부득이한 경우 MockBean을 쓰는 것은 편리할 것입니다. MockBean 의 단점은 실제 객체의 행위가 달라졌는데도 Mock은 변하지 않는다는 것입니다! 이점을 유념해주세요! 하지만 왠만하면 Mock은 테스트에 혼란을 주는 경우가 많습니다!

3. 이것은 팀원과 이야기하는 것이 가장 중요하지만 안될 거 없죠!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/TaskService.java:
```diff
@@ -1,50 +1,60 @@
+import java.util.TreeMap;
 
 @Service
 public class TaskService {
```
객체지향 5대 원리로 불리는 SOLID를  읽어보시고 
객체를 좀 더 나눠봐도 좋을 것 같습니다! 

일단 인터페이스를 만들어 보는 것부터 시작해보면 좋겠어요!

steve7867 replied:
역할에 따라 객체들을 분리해봤습니다.
기존의 TaskService를 인터페이스로 만들고 TaskServiceImpl을 구현체로 두었습니다.
그리고 task를 저장하는 책임은 TaskRepository 인터페이스로 위임했습니다.
task의 id를 할당하는 책임도 TaskIdGenerator로 위임했습니다.

steve7867 commented:
제가 이틀 전에 코드 리뷰 작성해 놓고 Submit Review를 안 눌러서 그 동안 작성한 코멘트들이 전부 pending 되었나 봅니다.
죄송합니다 ㅠㅠ

steve7867 committed:
getId, getTitle 메소드에 대한 테스트 코드 삭제

불필요한 테스트라 판단되어 삭제

steve7867 committed:
TaskService 인터페이스 분리

steve7867 committed:
TaskService 객체 분리

task 저장 역할을 하는 TaskRepository, Id 할당 역할을하는 TaskIdGenerator를 다른 객체로 분리

steve7867 committed:
TaskService 객체 분리에 따른 테스트 코드 수정

steve7867 committed:
DuplicateTaskException 추가 및 예외 처리 로직 추가

steve7867 committed:
Task: 필드 추가, 빌더 패턴 적용, 메소드 수정

등록일, 수정일, 마감일을 나타내는 필드 추가.
빌더 패턴 적용.
메소드 수정: 필드 추가에 따른 equals 메소드 수정, changeTitle, changeDeadLine 메소드 추가

steve7867 committed:
TaskDto 추가

steve7867 committed:
TaskDto 추가에 따른 이름 변경

steve7867 committed:
TaskDto 추가에 따른 리팩토링, 기능 추가

리팩토링: 기존의 Task 객체를 사용하던 코드를 TaskDto를 사용하도록 변경.
기능 추가: 마감일 순으로 task를 요청하는 api 추가

steve7867 commented:
오늘 변경된 코드와 추가된 기능이 조금 많습니다.

1. TaskService를 TaskService 인터페이스와 TaskServiceImpl 클래스로 나눔.
2. task를 저장하는 책임을 TaskRepository에 위임.
3. task의 id를 할당하는 책임을 TaskIdGenerator에 위임.
4. 클라이언트와의 통신에서 Task 객체 대신 TaskDto 객체를 사용하도록 변경.
5. Task 객체에 등록일, 수정일, 마감일을 나타내는 필드 추가.
6. task를 마감일 순으로 요청하는 api 추가 (GET /tasks/deadline)
7. 기존에 있던 task의 title과 똑같은 title을 가진 task의 추가 요청을 거부함.

변경된 코드에 대한 테스트 코드 작성은 아직 하지 못했습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/models/TaskDto.java:
```diff
@@ -0,0 +1,60 @@
+        this.deadLine = deadLine;
+    }
+
+    private TaskDto() {
+    }
```
기본생성자는 되도록 없애는 것이 좋습니다!

steve7867 replied:
기본 생성자를 삭제하고 모든 필드에 대한 파라미터를 가지는 생성자를 만들었습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/models/TaskDto.java:
```diff
@@ -0,0 +1,60 @@
+    public TaskDto(@JsonProperty String title, @JsonProperty LocalDateTime deadLine) {
+        this.title = title;
+
+        if (deadLine == null) {
+            return;
+        }
```
이 코드의 의미가 무엇인가요?

steve7867 replied:
불필요하네요.
삭제하겠습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/TaskServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+import java.util.stream.Collectors;
+
+@Service
+public class TaskServiceImpl implements TaskService {
```
인터페이스와 클래스를 나누셨군요! 하지만 SOLID 원칙과는 조금 다른 듯 합니다!

SOLID 원칙에 대해서 이해하고 있으신 내용을 정리하여 여기에 적어주세요!

steve7867 replied:
인터페이스와 클래스를 나누고, task의 id를 생성하는 책임을 TaskIdGenerator 객체에 위임했는데 인터페이스가 아니라 구현 객체에 의존하도록 코드를 작성했네요. 이는 개방 폐쇄 원칙과 의존 역전 원칙에 부합하지 않습니다. 그리고 Task와 TaskDto간의 변환 작업이 여전히 TaskServiceImpl 클래스에서 이루어지고 있으므로 단일 책임의 원칙도 제대로 지킨 것이라고 할 수 없습니다. 그래서 TaskIdGenerator 인터페이스를 만들고 TaskIdGeneratorImpl 구현 객체로 나누어 TaskService는 TaskIdGenerator 인터페이스에 의존하도록 변경했습니다. 그리고 Task, TaskDto간 변환 작업의 책임을 가지는 util 클래스인 TaskConverter 클래스를 만들어 여기에 변환 작업의 책임을 위임했습니다.

단일 책임의 원칙(SRP, Single Responsibility Principle)
- 하나의 모듈은 한 가지 책임을 가져야 한다.
- 모듈이 변경되는 이유는 한가지여야 한다.
- 변경이 필요할 때 대상이 명확해진다.

개방 폐쇠 원칙 (OCP, Open-Closed Principle)
- 확장에 대해서는 열려있고, 수정에 대해서는 닫혀있어야 한다.
  - 요구사항이 변경될 때, 새로운 동작을 추가하여 애플리케이션의 기능을 확장할 수 있어야 한다.
  - 기존의 코드를 수정하지 않고 애플리케이션의 동작을 추가하거나 변경할 수 있어야 한다.
- 개방 폐쇠 원칙을 지키기 위해서는 추상화에 의존해야 한다.
  - 추상화를 통해 변하지 않는 것만 남김으로써 기능을 구체화하고 확장할 수 있다.
  - 변하는 것들은 숨기고 변하지 않는 것들만 남긴다.

인터페이스 분리 원칙(ISP, Interface Segregation Principle)
- 클라이언트의 목적과 용도에 적합한 인터페이스만을 제공해야 한다.
- 클라이언트는 자신이 사용하지 않는 메소드에 의존해서는 안 된다.

리스코프 치환 원칙(LSP, Liskov Substitution Principle)
- 하위 타입은 상위 타입을 대체할 수 있어야 한다.
- 해당 객체를 사용하는 클라이언트는 상위 타입이 하위 타입으로 대체되어도 차이점을 인식하지 못한 채 상위 타입의 퍼블릭 인터페이스를 통해 서브 클래스를 사용할 수 있어야 한다.

의존 역전 원칙(DIP, Dependency Inversion Principle)
- 고수준 모듈은 저수준 모듈의 구현에 의존해서는 안 된다.
- 저수준 모듈이 고수준 모듈에서 정의한 추상 타입에 의존해야 한다.

어떤 부분이 SOLID 원칙에 위배되는지 알려주실 수 있을까요?

ssisksl77 replied:
일단 TaskService가 단일책임원칙에 위배되었습니다. 범위가 너무 큽니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/TaskServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    public TaskDto createTask(TaskDto dto) {
+        final Task task = dtoToTask(idGenerator.generateId(), dto);
+        final Task addedTask = taskRepository.addTask(task);
+
+        return taskToDto(addedTask);
```
Dto가 서비스의 영역이라고 설계를 하신 것으로 이해되는데요! 맞을까요?

steve7867 replied:
네 맞습니다.
컨트롤러에서 클라이언트와의 통신에 사용되며, serivce 계층에서 Task 객체로 변환되어 repository 계층으로 전달되도록 설계했습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -1,51 +1,71 @@
 
     @GetMapping
-    public List<Task> list() {
+    public Collection<TaskDto> collection() {
```
List와 Collection의 차이를 댓글로 공유해주세요!

steve7867 replied:
Collection의 경우 List, Set, Queue의 상위 인터페이스로써 Map을 구현하는 클래스를 제외한 나머지 컬렉션 프레임워크 클래스들의 최상위 인터페이스라고 할 수 있습니다. Collection의 경우 요소의 중복을 허용할 수도 있고 허용하지 않을 수도 있고, 요소의 순서를 보장할 수도 있고 보장하지 않을 수도 있습니다. List의 경우, 요소의 중복을 허용하고 순서를 보장한다는 점에서 차이가 있습니다. 반면 Set의 경우에는 중복을 불허하고 순서도 보장하지 않습니다. 물론 LinkedHashSet의 경우는 중복은 불허하지만 루핑할 때 순서를 보장해주는 Set도 있습니다.

steve7867 replied:
4일 전에 같은 말씀을 하셔서 댓글 달아 놓은 게 있습니다.
확인 부탁드립니다.

ssisksl77 replied:
넵 감사합니다! 확인하였습니다! 너무 잘 적으셨어요! 아래에서 잠깐 적어놓았지만! 

내용에 대한 공식 레퍼런스 링크 같은 것이 있는 것이 지식 공유에 많은 도움이 될거에요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -1,23 +1,125 @@
+        }
+
+        public Builder deadLine(LocalDateTime deadLine) {
+            if (deadLine == null) {
+                return this;
+            }
```
꼭 있어야 하는 코드인가요? 좀 더 설명해주시면 좋겠어요!

steve7867 replied:
클라이언트 쪽에서 deadLine을 입력하지 않은 경우 TaskDto 객체의 deadLine 필드에는 null 값이 들어가게 됩니다.
TaskDto 객체의 deadLine이 null일 때는 마감기한이 없는 것으로 간주하여 Task 객체의 deadLine 필드에는 LocalDateTime.MAX가 대입되도록 하기 위함입니다.

ssisksl77 replied:
아하! 좋습니다.

하지만 Builder 객체가 단일 책임과는 멀어지는 것이 아닌가 생각이 듭니다.
빌더는 Task를 이쁘게 만드는 것만으로도 벅차야 하는데,
Task의 디폴트값에 대한 로직까지 같고 있으니 말이죠! 
Task의 생성자가 맡아서 해야겠습니다.

steve7867 replied:
네 수정해보겠습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/utils/TaskDtoValidator.java:
```diff
@@ -0,0 +1,22 @@
+
+public class TaskDtoValidator {
+
+    public static void validate(TaskDto dto) {
+        if (dto == null) {
+            throw new NullTaskDtoException();
+        }
+
+        validateTitle(dto.getTitle());
+    }
+
+    private static void validateTitle(String title) {
+        if (title == null || title.isBlank()) {
+            throw new InvalidTaskDtoTitleException(title);
+        }
+    }
```
Dto가 null일 때와 title의 null일 때 서로 다른 예외가 있는 이유를 설명해주시면 감사하겠어요!

steve7867 replied:
생각해보니 dto가 null인 경우가 없어 보입니다.
NullTaskException은 삭제했습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepositoryImpl.java:
```diff
@@ -0,0 +1,94 @@
+@Repository
+public class TaskRepositoryImpl implements TaskRepository {
+
+    private final Map<Long, Task> tasks = Collections.synchronizedMap(new TreeMap<>());
```
ConcurrentHashMap 와 어떤 차이가 있는지 댓글로 공유해주세요!

steve7867 replied:
둘 다 멀티 스레드 환경에서 스레드 간의 동시성을 보장한다는 공통점이 있습니다. 
자바의 HashMap은 내부적으로 체이닝 방식으로 구현되어 있습니다.  기본적으로 사이즈가 16인 테이블이 있고 테이블 요소 하나하나를 버킷 또는 세그먼트라고 부릅니다. (여기서는 버킷이라 하겠습니다.) 이 버킷은 linkedlist로 되어 있으나 사이즈가 증가하게 될 경우 red black tree로 (더 정확히는 llrb) 바뀌게 되어있습니다.
ConcurrentHashMap은 write operation에 대해서만 버킷 단위로 locking을 하는 반면 Collections.synchronizedMap()은 read, write operation에 대해 테이블 전체에 locking을 합니다.
ConcurrentHashMap은 null key, null value를 허용하지 않지만 Collections.synchronizedMap()의 경우에는 인자로 주어지는 backing map의 type에 따라 경우가 나뉩니다. HashMap이나 LinkedHashMap이 인자로 주어지게 되면 하나의 null key, 다수의 null value를 허용하지만 TreeMap의 경우에는 null value는 허용하지만 null key는 허용하지 않습니다.
또한 ConcurrentHashMap은 하나의 스레드가 iterate 작업을 하는 중에 다른 스레드에 의해 map의 변형이 발생해도 ConcurrentModifiedException이 발생하지 않지만(fail-safe) Collections.synchronizedMap()은 backing map에 따라서 fail-fast, fail-safe가 나뉘는 특징이 있습니다.
따라서 성능 면에서는 ConcurrentHashMap이 유리하지만 데이터 일관성의 관점에서 보면 Collections.synchronizedMap()이 더 유리하다고 할 수 있습니다. (ConcurrentHashMap은 하나의 스레드가 read operation 중에 다른 스레드가 write operation을 수행할 수 있기 때문에 데이터의 일관성이 완벽하게 보장되지 않습니다.)

ssisksl77 replied:
좋습니다. 읽기의 완벽한 일관성을 위해 SynchronizedMap로 전부 락을 걸어버렸다! 라고 보면 되겠군요!

조금 아쉬운 점이 있다면 ConcurrentHashMap와 어떤 차이가 있는지에 대한 문의에서 다른 여러가지 내용이 산재된 것입니다.

1. A와 B는 서로 어떻게 다른지
2. 무엇이 더 중요하다고 생각하여 A를 선택했다.

라는 흐름과는 전혀다르게 공부하셨던 지식만을 나열하신 느낌입니다!
또한 이런 여러가지 지식을 나열하셨을 때에는 관련 공식레퍼런스의 링크정보가 있는 것이 좋습니다!

steve7867 replied:
말씀해주신 점 꼭 숙지하겠습니다.

steve7867 committed:
Task: 불필요한 로직 제거, TaskDto: 기본 생성자 제거, NullTaskDtoException 제거

steve7867 committed:
TaskDto: 기본 생성자 제거, NullTaskDtoException 제거

steve7867 committed:
Merge remote-tracking branch 'origin/feature/day1' into feature/day1

steve7867 committed:
Task: Builder 메소드 수정, TaskRepositoryImpl: Map 구현체 변경

Builder 메소드 수정: 클라이언트 측에서 마감 기한을 입력하지 않을 경우, Task 객체의 deadLine 필드에 LocalDateTime.MAX을 대입하기 위함
Map 구현체 변경: 저장된 Task 객체 간에 순서를 보장해줄 필요가 없으므로 TreeMap에서 HashMap으로 변경

steve7867 committed:
TaskIdGenerator, TaskConverter 추가

TaskService에서 Task 객체의 id를 생성하는 책임을 TaskIdGenerator에 위임하고 Task 객체와 TaskDto 객체 간의 변환 책임을 TaskConverter에 위임.

steve7867 committed:
Merge pull request #1 from steve7867/feature/day1

Task: 불필요한 로직 제거, TaskDto: 기본 생성자 제거, NullTaskDtoException 제거

steve7867 commented:
몇일 전에 제가 작성한 댓글들이 pending되어 그저께 다시 올라갔습니다.
확인하셨는지 잘 모르겠습니다.
확인하셨으면 확인 댓글 부탁드립니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/TaskService.java:
```diff
@@ -1,54 +1,37 @@
-public class TaskService {
-    private List<Task> tasks = new ArrayList<>();
-    private Long newId = 0L;
+public interface TaskService {
```
nit : 메소드 내 인터페이스는 필수입니다.

steve7867 replied:
조금만 더 상세하게 말씀해주실 수 있을까요?
어떤 의미인지 잘 모르겠습니다.

ssisksl77 replied:
아 휴... 제가 잘못적었습니다. 순간 컴퓨터가 맛이가서 답변이 늦었네요... 크흠...!!

메소드 내 주석은 필수입니다!!

steve7867 replied:
주석 작성하는 걸 습관화하겠습니다.
